core/jvm/src/main/scala/fs2/interop/flow/StreamSubscription.scala

```diff
@@ -76,95 +95,93 @@ private[flow] final class StreamSubscription[F[_], A] private (
         .compile
         .drain
 
+    val cancellation = F.asyncCheckAttempt[Unit] { cb =>
+      F.delay {
+        // Check if we were already cancelled before calling run.
+        if (!canceled.compareAndSet(StreamSubscription.Sentinel, () => cb.apply(Either.unit))) {
+          Either.unit
+        } else {
+          Left(Some(F.unit))
+        }
+      }
+    }
+
     events
-      .race(canceled.get)
+      .race(cancellation)
       .guaranteeCase {
         case Outcome.Succeeded(result) =>
           result.flatMap {
-            case Left(())  => onComplete // Events finished normally.
+            case Left(())  => F.delay(onComplete()) // Events finished normally.
             case Right(()) => F.unit // Events was canceled.
           }
-        case Outcome.Errored(ex) => onError(ex)
-        case Outcome.Canceled()  => cancelMe
+        case Outcome.Errored(ex) => F.delay(onError(ex))
+        case Outcome.Canceled() =>
+          F.delay(onError(new CancellationException("StreamSubscription.run was canceled")))
       }
       .void
   }
 
   // According to the spec, it's acceptable for a concurrent cancel to not
-  // be processed immediately, but if you have synchronous `cancel();
-  // request()`, then the request _must_ be a NOOP. Fortunately,
-  // ordering is guaranteed by a sequential dispatcher.
+  // be processed immediately, but if you have synchronous
+  // `cancel(); request()`,
+  // then the request must be a NOOP.
   // See https://github.com/zainab-ali/fs2-reactive-streams/issues/29
   // and https://github.com/zainab-ali/fs2-reactive-streams/issues/46
-  private def cancelMe: F[Unit] =
-    canceled.complete(()).void
-
-  override def cancel(): Unit =
-    try
-      requestDispatcher.unsafeRunAndForget(cancelMe)
-    catch {
-      case _: IllegalStateException =>
-      // Dispatcher already shutdown, we are on terminal state, NOOP.
+  override def cancel(): Unit = {
+    val cancelCB = canceled.getAndSet(null)
+    if (cancelCB ne null) {
+      cancelCB.apply()
     }
+  }
 
-  override def request(n: Long): Unit = {
-    val prog =
-      canceled.tryGet.flatMap {
-        case None =>
-          if (n == java.lang.Long.MAX_VALUE)
-            requests.offer(StreamSubscription.Request.Infinite)
-          else if (n > 0)
-            requests.offer(StreamSubscription.Request.Finite(n))
-          else
-            onError(
-              ex = new IllegalArgumentException(s"Invalid number of elements [${n}]")
-            )
+  override def request(n: Long): Unit =
+    if (canceled.get() ne null) {
```

Comment:
Is this `canceled` check really doing anything useful?

Reply:
Oh, yes it is, this comment would almost be more helpful here.
https://github.com/typelevel/fs2/blob/1f6bc986c1c2233a53d044c8462269d80380f771/core/jvm/src/main/scala/fs2/interop/flow/StreamSubscription.scala#L124-L129

Reply:
Well, that was the original place of the comment, I just polished it a little.
I can move it if you prefer, the thing is that it also affects `cancel`, and well both methods are close together.